File: code/rust/ch06/src/vm/instruction.rs
use vm::opcodes::OPCODES;

const MAXARG_BX: i32 = 1 << 18 - 1; // 262143
const MAXARG_SBX: i32 = MAXARG_BX >> 1; // 131071

/*
 31       22       13       5    0
  +-------+^------+-^-----+-^-----
  |b=9bits |c=9bits |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |    bx=18bits    |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |   sbx=18bits    |a=8bits|op=6|
  +-------+^------+-^-----+-^-----
  |    ax=26bits            |op=6|
  +-------+^------+-^-----+-^-----
 31      23      15       7      0
*/
pub trait Instruction {
    fn opname(self) -> &'static str;
    fn opmode(self) -> u8;
    fn b_mode(self) -> u8;
    fn c_mode(self) -> u8;
    fn opcode(self) -> i32;
    fn abc(self) -> (i32, i32, i32);
    fn a_bx(self) -> (i32, i32);
    fn a_sbx(self) -> (i32, i32);
    fn ax(self) -> i32;
}

impl Instruction for u32 {
    fn opname(self) -> &'static str {
        OPCODES[self.opcode() as usize].name
    }

    fn opmode(self) -> u8 {
        OPCODES[self.opcode() as usize].opmode
    }

    fn b_mode(self) -> u8 {
        OPCODES[self.opcode() as usize].bmode
    }

    fn c_mode(self) -> u8 {
        OPCODES[self.opcode() as usize].cmode
    }

    fn opcode(self) -> i32 {
        self as i32 & 0x3F
    }

    fn abc(self) -> (i32, i32, i32) {
        let a = (self >> 6 & 0xFF) as i32;
        let c = (self >> 14 & 0x1FF) as i32;
        let b = (self >> 23 & 0x1FF) as i32;
        (a, b, c)
    }

    fn a_bx(self) -> (i32, i32) {
        let a = (self >> 6 & 0xFF) as i32;
        let bx = (self >> 14) as i32;
        (a, bx)
    }

    fn a_sbx(self) -> (i32, i32) {
        let (a, bx) = self.a_bx();
        (a, bx - MAXARG_SBX)
    }

    fn ax(self) -> i32 {
        (self >> 6) as i32
    }
}

File: code/csharp/ch12/luavm/vm/Instruction.cs
using System;
using luavm.api;
using LuaVM = luavm.state.LuaState;

namespace luavm.vm
{
    public class Instruction
    {
        public Instruction(uint b)
        {
            _self = b;
        }

        private readonly uint _self;

        public int Opcode()
        {
            return (int) (_self & 0x3F);
        }

        public (int, int, int) ABC()
        {
            return (
                (int) (_self >> 6) & 0xFF,
                (int) (_self >> 23) & 0x1FF,
                (int) (_self >> 14) & 0x1FF
            );
        }

        public (int, int) ABx()
        {
            return ((int) ((_self >> 6) & 0xFF),
                (int) (_self >> 14));
        }

        private const int MAXARG_Bx = (1 << 18) - 1;
        private const int MAXARG_sBx = MAXARG_Bx >> 1;

        public (int, int) AsBx()
        {
            var (a, sBx) = ABx();
            return (a, sBx - MAXARG_sBx);
        }

        public int Ax()
        {
            return (int) (_self >> 6);
        }

        private string OpName()
        {
            return Constant.Opcodes[Opcode()].Name;
        }

        public byte OpMode()
        {
            return Constant.Opcodes[Opcode()].OpMode;
        }

        public byte BMode()
        {
            return Constant.Opcodes[Opcode()].ArgBMode;
        }

        public byte CMode()
        {
            return Constant.Opcodes[Opcode()].ArgCMode;
        }

        private static void printStack(LuaState ls)
        {
            var top = ls.GetTop();
            for (var i = 1; i <= top; i++)
            {
                var t = ls.Type(i);
                switch (t)
                {
                    case Constant.LUA_TBOOLEAN:
                        Console.Write($"[{ls.ToBoolean(i)}]");
                        break;
                    case Constant.LUA_TNUMBER:
                        Console.Write($"[{ls.ToNumber(i)}]");
                        break;
                    case Constant.LUA_TSTRING:
                        Console.Write($"[{ls.ToString(i)}]");
                        break;
                    default:
                        Console.Write($"[{ls.TypeName(t)}]");
                        break;
                }
            }

            Console.WriteLine();
        }

        public void Execute(Opcode op, LuaVM vm)
        {
            var action = op.Action;

            if (action != null)
            {
                action(this, ref vm);
            }
            else
            {
                throw new Exception(OpName());
            }
        }
    }
}